app/models/anonymous_user.rb

```diff
@@ -49,7 +49,7 @@ def name(*_args); I18n.t(:label_user_anonymous) end
 
   def mail; nil end
 
-  def time_zone; nil end
+  def time_zone; ActiveSupport::TimeZone[Setting.user_default_timezone.presence || "Etc/UTC"] end
```

Comment:
After reading your comment I thought, "right he is". Then I tried it out:

```
pry(main)> User.system.build_preference
  SystemUser Load (0.5ms)  SELECT "users".* FROM "users" WHERE "users"."type" = $1 ORDER BY "users"."id" ASC LIMIT $2  [["type", "SystemUser"], ["LIMIT", 1]]
  UserPreference Load (0.2ms)  SELECT "user_preferences".* FROM "user_preferences" WHERE "user_preferences"."user_id" = $1 LIMIT $2  [["user_id", 1], ["LIMIT", 1]]
=> #<UserPreference:0x000000030eaf0148 id: nil, user_id: 1, settings: {}, created_at: Mon, 08 Jul 2024 07:17:52.591795000 UTC +00:00, updated_at: Mon, 08 Jul 2024 07:17:52.591795000 UTC +00:00>
```

That happens because the user objects are persisted. It wouldn't happen on unpersisted models. 

I think this would not amount to much additional time per request when done but on the other hand the code necessary is quite easy to maintain so I would keep it as it is. 

An improvement I see, though is to extract code shared between the three internal user objects into a module.

What do you think @cbliard?

Reply:
Nice find.

> I think this would not amount to much additional time per request when done but on the other hand the code necessary is quite easy to maintain so I would keep it as it is.

Yeah, I was not really into the time it would take, but rather the logic duplication in the 3 classes, and in the fallback to Etc/UTC with a regular user.

> An improvement I see, though is to extract code shared between the three internal user objects into a module.

Yep 👍

Reply:
Did that in [cd615f5](https://github.com/opf/openproject/pull/16567/commits/cd615f5108034852b1473eb775e55574951163bb).